src/urlpattern_constructor_string_parser.cpp

```diff
@@ -0,0 +1,579 @@
+  //   previous token’s type is "regexp".
+  //   previous token’s type is "close".
+  //   previous token’s type is "asterisk".
+
+  if (prev_token->type == TOKEN_TYPE::NAME ||
+      prev_token->type == TOKEN_TYPE::REGEXP ||
+      prev_token->type == TOKEN_TYPE::CLOSE ||
+      prev_token->type == TOKEN_TYPE::ASTERISK) {
+    return false;
+  }
+
+  return true;
+}
+
+// https://wicg.github.io/urlpattern/#next-is-authority-slashes
+ada_really_inline bool constructor_string_parser::next_is_authority_slashes() {
+  // 1. If the result of running is a non-special pattern char given parser,
+  // parser’s token index + 1, and "/" is false, then return false.
+  if (!is_nonspecial_pattern_char(token_index + 1, U"/")) {
+    return false;
+  }
+
+  // 2. If the result of running is a non-special pattern char given parser,
+  // parser’s token index + 2, and "/" is false, then return false.
+  if (!is_nonspecial_pattern_char(token_index + 2, U"/")) {
+    return false;
+  }
+
+  return true;
+}
+
+// https://wicg.github.io/urlpattern/#is-a-non-special-pattern-char
+ada_really_inline bool constructor_string_parser::is_nonspecial_pattern_char(
+    size_t index, const char32_t *value) {
+  // 1. Let token be the result of running get a safe token given parser and
+  // index.
+  token *safe_token = get_safe_token(index);
+
+  // 2. If token’s value is not value, then return false.
+  if (safe_token->value.compare(std::u32string_view(value)) != 0) {
+    return false;
+  }
+
+  // If any of the following are true:
+  //
+  //   token’s type is "char";
+  //   token’s type is "escaped-char"; or
+  //   token’s type is "invalid-char",
+  //
+  // then return true.
+
+  if (safe_token->type == TOKEN_TYPE::CHAR ||
+      safe_token->type == TOKEN_TYPE::ESCAPED_CHAR ||
+      safe_token->type == TOKEN_TYPE::INVALID_CHAR) {
+    return true;
+  }
+
+  return false;
+}
+
+// https://wicg.github.io/urlpattern/#get-a-safe-token
+ada_really_inline token *constructor_string_parser::get_safe_token(
+    size_t &index) {
+  // 1. If index is less than parser’s token list's size, then return parser’s
+  // token list[index].
+  if (index < token_list.size()) return &token_list[index];
+  // 2. Assert: parser’s token list's size is greater than or equal to 1.
+  // TODO: messages for the asserts? conditional and then throw error?
+  assert(token_list.size() >= 1);
+  // 3. Let last index be parser’s token list's size − 1.
+  size_t last_index = token_list.size() - 1;
+  // 4. Let token be parser’s token list[last index].
+  // 5. Assert: token’s type is "end".
+  assert(token_list[last_index].type == TOKEN_TYPE::END);
+  return &token_list[last_index];
+}
+
+// https://wicg.github.io/urlpattern/#is-a-group-open
+ada_really_inline bool constructor_string_parser::is_group_open() {
+  // If parser’s token list[parser’s token index]'s type is "open", then
+  // return true. Else return false.
+  return token_list[token_index].type == TOKEN_TYPE::OPEN;
```

Comment:
This is not safe. We should add development asserts to make sure token_index is smaller than token_list length.